File: app/src/main/java/com/repelliuss/npuzzle/utils/Move.java
package com.repelliuss.npuzzle.utils;

public enum Move {

    STAY,
    LEFT,
    RIGHT,
    UP,
    DOWN;

    public static Move fromAngle(double angle) {
        if(inRange(angle, 45, 135)) {
            return UP;
        }
        else if(inRange(angle, 0,45) || inRange(angle, 315, 360)) {
            return RIGHT;
        }
        else if(inRange(angle, 225, 315)) {
            return DOWN;
        }
        else {
            return LEFT;
        }

    }

    private static boolean inRange(double angle, float init, float end){
        return (angle >= init) && (angle < end);
    }

    public static void swap(Move[] arr, int left, int right) {
        Move temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
    }
}

File: deep_learning/game/move.py
#!/usr/bin/env python3

from enum import Enum

class Move(Enum):
    STAY = 0
    LEFT = 1
    RIGHT = 2
    UP = 3
    DOWN = 4

    def to_int(move):
        if move == Move.LEFT:
            return 0
        elif move == Move.RIGHT:
            return 1
        elif move == Move.UP:
            return 2
        elif move == Move.DOWN:
            return 3
        else:
            return 0

    def opposite(move):
        if move == Move.LEFT:
            return Move.RIGHT
        elif move == Move.RIGHT:
            return Move.LEFT
        elif move == Move.UP:
            return Move.DOWN
        elif move == Move.DOWN:
            return Move.UP
        else:
            return None
